tools/server/server.cpp

```diff
@@ -2188,6 +2188,24 @@ struct server_context {
 
         metrics.init();
 
+        // thinking is enabled if:
+        // 1. It's not explicitly disabled (reasoning_budget == 0)
+        // 2. The chat template supports it
+        bool enable_thinking = params_base.reasoning_budget != 0;
+        if (enable_thinking) {
+            common_chat_templates_inputs dummy_inputs;
+            common_chat_msg msg;
+            msg.role = "user";
+            msg.content = "test";
+            dummy_inputs.messages = {msg};
+            dummy_inputs.enable_thinking = false;
+            const auto rendered_no_thinking = common_chat_templates_apply(chat_templates.get(), dummy_inputs);
+            dummy_inputs.enable_thinking = true;
+            const auto rendered_with_thinking = common_chat_templates_apply(chat_templates.get(), dummy_inputs);
+            enable_thinking = rendered_no_thinking.prompt != rendered_with_thinking.prompt;
```

Comment:
I don't understand the logic of this check - seems we would disable thinking if the 2 rendered prompts are different. Why would that be?

Reply:
I'll double check in a few, but I think the logic as written is that if the two prompts are different (with and without thinking), the `enable_thinking` bool is set to `true`. If they match, the assumption is that the `enable_thinking` flag is ignored in the template. This is certainly a heuristic since some templates might only honor it under certain conditions (eg the funky Granite 3.x XOR conditions with `documents` and `tools`).